pandas/core/algorithms.py

```diff
@@ -1537,40 +1538,79 @@ def _take_nd_object(arr, indexer, out, axis: int, fill_value, mask_info):
+    Part of _get_take_nd_function below that doesn't need `mask_info` and thus
+    can be cached (mask_info potentially contains a numpy ndarray which is not
+    hashable and thus cannot be used as argument for cached function).
+    """
+    tup = (arr_dtype.name, out_dtype.name)
+    if ndim == 1:
+        func = _take_1d_dict.get(tup, None)
+    elif ndim == 2:
+        if axis == 0:
+            func = _take_2d_axis0_dict.get(tup, None)
+        else:
+            func = _take_2d_axis1_dict.get(tup, None)
+    if func is not None:
+        return func
+
+    tup = (out_dtype.name, out_dtype.name)
+    if ndim == 1:
+        func = _take_1d_dict.get(tup, None)
+    elif ndim == 2:
+        if axis == 0:
+            func = _take_2d_axis0_dict.get(tup, None)
+        else:
+            func = _take_2d_axis1_dict.get(tup, None)
+    if func is not None:
+        func = _convert_wrapper(func, out_dtype)
+        return func
+
+    return None
+
+
 def _get_take_nd_function(
     ndim: int, arr_dtype, out_dtype, axis: int = 0, mask_info=None
 ):
+    """
+    Get the appropriate "take" implementation for the given dimension, axis
+    and dtypes.
+    """
+    func = None
     if ndim <= 2:
-        tup = (arr_dtype.name, out_dtype.name)
-        if ndim == 1:
-            func = _take_1d_dict.get(tup, None)
-        elif ndim == 2:
-            if axis == 0:
-                func = _take_2d_axis0_dict.get(tup, None)
-            else:
-                func = _take_2d_axis1_dict.get(tup, None)
-        if func is not None:
-            return func
-
-        tup = (out_dtype.name, out_dtype.name)
-        if ndim == 1:
-            func = _take_1d_dict.get(tup, None)
-        elif ndim == 2:
-            if axis == 0:
-                func = _take_2d_axis0_dict.get(tup, None)
-            else:
-                func = _take_2d_axis1_dict.get(tup, None)
-        if func is not None:
-            func = _convert_wrapper(func, out_dtype)
-            return func
+        # for this part we don't need `mask_info` -> use the cached algo lookup
+        func = __get_take_nd_function_cached(ndim, arr_dtype, out_dtype, axis)
 
-    def func2(arr, indexer, out, fill_value=np.nan):
-        indexer = ensure_int64(indexer)
-        _take_nd_object(
-            arr, indexer, out, axis=axis, fill_value=fill_value, mask_info=mask_info
-        )
+    if func is None:
+
+        def func(arr, indexer, out, fill_value=np.nan):
+            indexer = ensure_int64(indexer)
+            _take_nd_object(
+                arr, indexer, out, axis=axis, fill_value=fill_value, mask_info=mask_info
+            )
+
+    return func
 
-    return func2
+
+@functools.lru_cache(maxsize=128)
+def _maybe_promote_cached(dtype, fill_value, fill_value_type):
+    # also use fill_value_type as (unused) argument to use this in the cache
+    # lookup -> differentiate 1 and True
+    return maybe_promote(dtype, fill_value)
+
+
+def _maybe_promote(dtype, fill_value):
+    try:
```

Comment:
this is very strange to do. can you simply change all of the call of `_maybe_promote` to `_maybe_promote_cached` that need it (e.g. in the code and not the tests).

putting a try/except *inside* here is reversing the paradigm and not good.

Reply:
That is not simply possible. We don't know in advance if the fill_value is going to be hashable or not. So that's the reason the fallback is needed.

Reply:
sure but the try/except needs to be in the cached method, NOT here. IOW you are now exposing 2 api's, we need to have exactly one.

Reply:
But try/except is because of non-hashable fill_values, which thus cannot be *inside* the cached method, that's the whole reason I added the try/except in the first place. 

I am not exposing two different APIs. These are internal helper methods, and `_maybe_promote_cached` is only used for `_maybe_promote`, and it is `_maybe_promote` that is then used internally in the `take` implementation. 
I can add a comment to `_maybe_promote_cached` that this is only the cached part of `_maybe_promote` and not to be used elsewhere.

Reply:
can we just simply ban non-hashables from `_maybe_promote`?